src/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -50,13 +105,121 @@ rawPageNumBlobs :: RawPage -> Word16
 rawPageNumBlobs (RawPage off ba) = indexByteArray ba (off + 1)
 
 type KeyOffset = Word16
+type ValueOffset = Word16
 
-rawPageKeyOffsets :: RawPage -> Vector KeyOffset
+rawPageKeyOffsets :: RawPage -> P.Vector KeyOffset
 rawPageKeyOffsets page@(RawPage off ba) =
-    Vector (off + fromIntegral (div2 dirOffset)) (fromIntegral dirNumKeys) ba
+    P.Vector (off + fromIntegral (div2 dirOffset)) (fromIntegral dirNumKeys + 1) ba
   where
     Dir {..} = rawPageDirectory page
 
+-- | for non-single key page case
+rawPageValueOffsets :: RawPage -> P.Vector ValueOffset
+rawPageValueOffsets page@(RawPage off ba) =
+    P.Vector (off + fromIntegral (div2 dirOffset) + fromIntegral dirNumKeys) (fromIntegral dirNumKeys + 1) ba
+  where
+    Dir {..} = rawPageDirectory page
+
+-- | single key page case
+rawPageValueOffsets1 :: RawPage -> (Word16, Word32)
+rawPageValueOffsets1 page@(RawPage off ba) =
+    ( indexByteArray ba (off + fromIntegral (div2 dirOffset) + 1)
+    , indexByteArray ba (div2 (off + fromIntegral (div2 dirOffset)) + 1)
+    )
+  where
+    Dir {..} = rawPageDirectory page
+
+rawPageHasBlobRefAt :: RawPage -> Int -> Word64
+rawPageHasBlobRefAt _page@(RawPage off ba) i = do
+    let j = unsafeShiftR i 6 -- `div` 64
+    let k = i .&. 63         -- `mod` 64
+    let word = indexByteArray ba (div4 off + 1 + j)
+    unsafeShiftR word k .&. 1
+
+rawPageOpAt :: RawPage -> Int -> Word64
+rawPageOpAt page@(RawPage off ba) i = do
+    let j = unsafeShiftR i 5 -- `div` 32
+    let k = i .&. 31         -- `mod` 32
+    let word = indexByteArray ba (div4 off + 1 + roundUpTo64 (fromIntegral dirNumKeys) + j)
+    unsafeShiftR word (mul2 k) .&. 3
+  where
+    Dir {..} = rawPageDirectory page
+
+roundUpTo64 :: Int -> Int
+roundUpTo64 i =
+    if k == 0 then j else j + 1
+  where
+    j = unsafeShiftR i 6 -- `div` 64
+    k = i .&. 63         -- `mod` 64
```

Comment:
IIRC, this can be done without a branch like this:
```suggestion
roundUpTo64 :: Int -> Int
roundUpTo64 i = ((i+63) `unsafeShiftR` 6) .&. 63
```
(only lightly tested)

Reply:
For the record this `bitvec` has branches: https://github.com/Bodigrim/bitvec/blob/master/src/Data/Bit/Utils.hs#L80

And your implementation does something else:

```
*Database.LSMTree.Internal.RawPage> roundUpTo64 12345678
192902
*Database.LSMTree.Internal.RawPage> let duncan_roundUpTo64 i = ((i+63) `unsafeShiftR` 6) .&. 63
*Database.LSMTree.Internal.RawPage> duncan_roundUpTo64 (12345678 :: Int)
6
```

Reply:
Yes, sorry, I extracted it from code where I was also looking at the offset within a word.

So it's even simpler:
```
reference :: Int -> Int
reference x = if m==0 then d else d+1
  where
    (d,m) = divMod x 64

impl :: Int -> Int
impl = (i+63) `unsafeShiftR` 6

ok :: Bool
ok = and [ impl x == reference x | x <- [0..2^16+1] ]
```